### PR TITLE
Add earlyChecksumMode, with automatic patching of legacy checksums

### LINKED ISSUE #91: Change the checksum computation to be before parameter replacement
So, I have the plan that migration generation can generate `idx<platform>.migrations` files. These files contain the checksum and migration resource (this actually already exists in 12.11.0 in ebean).

The problem is that this checksum is based on the sql file resources as they exist on the file system. The checksum that ebean-migration uses is computed **after** the parameter expression replacement ( like `${...}` ). 

Ideally ebean-migration changes such that the checksum is computed **before** the parameter expression and in this way we can use  `idx<platform>.migrations` files to:

- avoid classpath scanning search for all the migrations
- compute the checksum at that point

That is, especially in CI/CD environment a lot of application starts happen with no actual migrations to run. ebean-migration does a pretty decent amount of work which gets more and more as the number of database migrations grows.  We can use the `idx<platform>.migrations` files to make this case a lot faster and more efficient which will be pretty handy when there are lots of DB migrations and/or those migration files are large.

The challenge is to more the computation of checksum (which will change the checksum value for migrations that have property evaluation / replacement) and do this transparently without breaking peoples existing migrations.

My current plan is to use the 0 migration which has a checksum value of 0.  Changing that to `1` to indicate that the new checksum is being used - when that 0 migration has a checksum value of `0` that would indicate that it uses the old checksum compute.

So we need to be careful here but we should get a good long term payoff in dealing with lots of migrations and or large migration files.  

FYI @rPraml

### PR DESCRIPTION
**Pull request by rob-bygrave**

### CODE REVIEW

File: ebean-migration/src/main/java/io/ebean/migration/runner/MigrationTable.java [hunk 1]
```diff
@@ -501,6 +526,10 @@ List<MigrationResource> runAll(List<LocalMigrationResource> localVersions) throw
         break;
       }
     }
+    if (patchLegacyChecksums && !checkStateOnly) {
+      // only patch the legacy checksums once
+      initMetaRow.resetChecksum(EARLY_MODE_CHECKSUM, connection, updateChecksumSql);
```
Comment (rPraml) [inline ebean-migration/src/main/java/io/ebean/migration/runner/MigrationTable.java:531]: CHECKME: Could this affect logicalLock? Or should this be done at the end, after all migrations are ran?

**Review by rPraml — COMMENTED** [review]
Yes, I like the idea with auto detection. It generally looks good for me.

I could imagine that there will be issues, if one of the migrations run into an error, then the <init> migration may contain EARLY_MODE_CHECKSUM and recovering from such a situation would be more difficult. (But to be honest, fortunately we have this case extremely rarely)

The next thing I didn't check in detail, if this change can interfer with the LogicalLock that stores the info in the same row. (I don't think so, because logicalLock just modifies the mcomment column)

Thanks for implementing this feature. I hope, we can test this soon here.

Roland

File: ebean-migration/src/main/java/io/ebean/migration/runner/MigrationTable.java (hunk 1 above)
Comment (rbygrave) [reply ebean-migration/src/main/java/io/ebean/migration/runner/MigrationTable.java:531]: It's all good with the logical lock, different columns being updated.  Note that the tests use H2 which uses the logicalLock. So yes, I'm happy that we are good with this.

**Comment by rbygrave** [comment]
> if one of the migrations run into an error, then the migration may contain EARLY_MODE_CHECKSUM and recovering from such a situation would be more difficult.

Well, there would have to be multiple commits or implied commits _AND_ there would have to be the desire to revert the checksums and go back to using "late mode checksums". Maybe we could make it easier to reset checksums though ... via a `patchResetChecksumVersions=*` or something like that (so with `*` as the versions, then patch the checksum on any versions that are mismatched).

**Pull request merged**